### PR TITLE
[RLlib] Implemented ViewRequirementConnector

### PR DESCRIPTION
**Pull request by kouroshHakha**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/connectors/tests/test_agent.py [hunk 1]
```diff
@@ -194,14 +194,13 @@ def test_view_requirement_connector(self):
...
 
 
-@pytest.mark.skip(reason="activate when view_requirement is fully implemented.")
 class TestViewRequirementConnector(unittest.TestCase):
     def test_vr_connector_respects_training_or_inference_vr_flags(self):
         """Tests that the connector respects the flags within view_requirements (i.e.
         used_for_training, used_for_compute_actions) under different is_training modes.
-        For inference,
+        is_training = False
             the returned data should be state -> obs
```
Comment (gjoliver) [inline rllib/connectors/tests/test_agent.py:167]: since we are here, can you help update this comment to be:
``the returned data is a SampleBatch that can be used to run corresponding policy```
and below should be:
```the returned data is the input dict itself, which the policy collector in env_runner will use to construct the episode, plus a SampleBatch that can be used to run corresponding policy```

File: rllib/connectors/tests/test_agent.py [hunk 2]
```diff
@@ -274,105 +276,85 @@ def test_vr_connector_shift_by_one(self):
...
+            # add cur obs to the list
+            obs_list.append(obs)
 
-    def test_vr_connector_causal_slice(self):
-        """Test that the ViewRequirementConnector can handle slice shifts correctly.
+            if t == 0:
+                check(for_action["prev_state"], for_action["state"])
+            else:
+                # prev state should be equal to the prev time step obs
+                check(for_action["prev_state"], obs_list[-2][None])
```
Comment (gjoliver) [inline rllib/connectors/tests/test_agent.py:266]: curious, what is this None key?

File: rllib/connectors/tests/test_agent.py [hunk 3]
```diff
@@ -274,105 +276,85 @@ def test_vr_connector_shift_by_one(self):
...
+            # check prev_strided_states_even
+            check(
+                for_action["prev_strided_states_even"],
+                np.stack(obs_list)[np.array([-5, -3, -1])][None],
+            )
+
+            check(
+                for_action["prev_strided_states_odd"],
+                np.stack(obs_list)[np.array([-4, -2])][None],
+            )
```
Comment (gjoliver) [inline rllib/connectors/tests/test_agent.py:322]: nice test!

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 4]
```diff
@@ -191,7 +215,101 @@ def add_action_reward_next_obs(self, values: Dict[str, TensorType]) -> None:
                     sub_list.append(flattened[i])
         self.agent_steps += 1
 
-    def build(self, view_requirements: ViewRequirementsDict) -> SampleBatch:
+    def build_for_inference(self) -> SampleBatch:
+        """During inference we will a samplebatch with a batch size of 1.
+        This data will only include the data for the last recorded timestep.
```
Comment (gjoliver) [inline rllib/evaluation/collectors/simple_list_collector.py]: this comment may be improved?
it may not only include data from the last timestep though.
just say ```During inference, we will build a SampleBatch with a batch size of 1 that can then be used to run the forward pass of a policy```?

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 5]
```diff
@@ -191,7 +215,101 @@ def add_action_reward_next_obs(self, values: Dict[str, TensorType]) -> None:
...
+        np_data = {}
+        for view_col, view_req in self.view_requirements.items():
+            # Create the batch of data from the different buffers.
+            data_col = view_req.data_col or view_col
+
+            # if this view is not for inference, skip it.
+            if not view_req.used_for_compute_actions:
+                continue
+            else:
+                if np.any(view_req.shift_arr > 0):
```
Comment (gjoliver) [inline rllib/evaluation/collectors/simple_list_collector.py]: you can actually take this check out of else. one fewer level of indentation.

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 6]
```diff
@@ -191,7 +215,101 @@ def add_action_reward_next_obs(self, values: Dict[str, TensorType]) -> None:
...
+
+            # Keep an np-array cache so we don't have to regenerate the
+            # np-array for different view_cols using to the same data_col.
+            if data_col not in np_data:
+                np_data[data_col] = [
+                    _to_float_np_array(d) for d in self.buffers[data_col]
+                ]
+
+            data = []
+            for d in np_data[data_col]:
```
Comment (gjoliver) [inline rllib/evaluation/collectors/simple_list_collector.py]: I am curious, can any of these be shared with build()?
I kind of feel like these should be same logics? just with different set of VRs?

File: rllib/evaluation/collectors/simple_list_collector.py [hunk 7]
```diff
@@ -217,10 +334,21 @@ def build(self, view_requirements: ViewRequirementsDict) -> SampleBatch:
             # Create the batch of data from the different buffers.
             data_col = view_req.data_col or view_col
 
-            # Some columns don't exist yet (get created during postprocessing).
-            # -> skip.
             if data_col not in self.buffers:
-                continue
+                # special treatment for state_out_<i>
+                # add them to the buffer in case they don't exist yet
+                if data_col.startswith("state_out_"):
```
Comment (gjoliver) [inline rllib/evaluation/collectors/simple_list_collector.py]: if this can be shared with build_for_inference(), then we don't need to have this block of duplicated code.

File: rllib/policy/policy.py [hunk 8]
```diff
@@ -1202,27 +1204,41 @@ def _get_dummy_batch_from_view_requirements(
...
+                    time_size = (
+                        len(view_req.shift_arr) if len(view_req.shift_arr) > 1 else None
+                    )
                     ret[view_col] = get_dummy_batch_for_space(
+                        view_req.space, batch_size=batch_size, time_size=time_size
+                    )
+                else:
+                    ret[view_col] = [view_req.space for _ in range(batch_size)]
+
+                # TODO: @kourosh remove these comments after this is tested
```
Comment (gjoliver) [inline rllib/policy/policy.py:1217]: please :)

**Review by gjoliver — CHANGES_REQUESTED** [review]
I am actually really excited about this change.
at the beginning, we didn't really know how to use agent connectors to build episodes for training. I feel like this PR paved the road for that, and it will make the EnvRunnerV2 logic much cleaner.
nice work. and aside from the couple of high level issues we discussed offline, I just have a few additional questions / minor comments.
thanks.

File: rllib/connectors/agent/view_requirement.py [hunk 9]
```diff
@@ -61,6 +78,9 @@ def transform(self, ac_data: AgentConnectorDataType) -> AgentConnectorDataType:
 
         env_id = ac_data.env_id
         agent_id = ac_data.agent_id
+        # TODO: we don't keep episode_id around so use env_id as episode_id ?
+        episode_id = env_id if SampleBatch.EPS_ID not in d else d[SampleBatch.EPS_ID]
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/connectors/agent/view_requirement.py:82]: yes. I think during sampling it gets generated. I don't know how the sampler makes sure it's unique tho. For example within one worker you can have a single env and a single agent but multiple episodes that are differentiated by the episode_id. So I think we should keep that differentiation here as well. This probably will show up when we use connector_enabled=True.

File: rllib/connectors/agent/view_requirement.py [hunk 10]
```diff
@@ -77,40 +97,22 @@ def transform(self, ac_data: AgentConnectorDataType) -> AgentConnectorDataType:
...
-            elif type(req.shift) == int:
-                h = req.shift
-            assert h <= 0, "Cannot use future data to compute action"
-            agent_batch[col] = agent_batch[col][h:]
-
-        sample_batch = self._get_sample_batch_for_action(vr, agent_batch)
+        agent_collector = self.agent_collectors[env_id][agent_id]
+
+        if SampleBatch.NEXT_OBS not in d:
+            raise ValueError(f"connector data {d} should contain next_obs.")
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/connectors/agent/view_requirement.py:103]: Yes it's baked into AgentCollector. But the logic there is absurd. new_obs should be there but then it gets removed from d and gets copied to obs keyword.

[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/connectors/agent/view_requirement.py:103]: I guess we can change this logic in the next PR.

File: rllib/connectors/tests/test_agent.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/connectors/tests/test_agent.py:167]: done.

File: rllib/connectors/tests/test_agent.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/connectors/tests/test_agent.py:266]: to add a single batch to the front of it. i.e if `obs_list[-2].shape = (1,)` then `obs_list[-2][None].shape = (1,1)`

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 4 above)
Comment (kouroshHakha) [reply rllib/evaluation/collectors/simple_list_collector.py]: done.

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 5 above)
Comment (kouroshHakha) [reply rllib/evaluation/collectors/simple_list_collector.py]: sg

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 6 above)
Comment (kouroshHakha) [reply rllib/evaluation/collectors/simple_list_collector.py]: ok shared most of the logic now. But the rest that are not shared have subtle differences that cannot be shared so easily.

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 7 above)
Comment (kouroshHakha) [reply rllib/evaluation/collectors/simple_list_collector.py]: done.

File: rllib/policy/policy.py (hunk 8 above)
Comment (kouroshHakha) [reply rllib/policy/policy.py]: done.

File: rllib/connectors/agent/view_requirement.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/connectors/agent/view_requirement.py:103]: yeah, sounds good.

File: rllib/connectors/tests/test_agent.py (hunk 2 above)
Comment (gjoliver) [reply rllib/connectors/tests/test_agent.py:268]: wow, didn't know you can do that. 👍

File: rllib/connectors/tests/test_agent.py [hunk 11]
```diff
@@ -381,44 +330,133 @@ def test_vr_connector_with_multiple_buffers(self):
...
+                SampleBatch.T: t - 1,
+            }
+            data = AgentConnectorDataType(0, 1, timestep_data)
+            processed = c([data])
+            for_action = processed[0].data.for_action
+
+            if t == 0:
+                obs_list.extend([obs_arrs[0] for _ in range(context_len)])
+                act_list.extend(
+                    [np.zeros_like(act_arrs[0]) for _ in range(context_len)]
```
Comment (gjoliver) [inline rllib/connectors/tests/test_agent.py:372]: just a nit picking TODO, maybe we should come back and make these actions different per time step.
don't know if that will reveal some deep bugs. right now everything is 0, things will match as long as the shapes match.
you don't need to add a TODO for this. if tests are already good.

File: rllib/connectors/tests/test_agent.py [hunk 12]
```diff
@@ -381,44 +330,133 @@ def test_vr_connector_with_multiple_buffers(self):
...
-                    check(for_action["context_act"], act_arrs[i - context_len : i])
+            else:
+                obs_list.pop(0)
+                act_list.pop(0)
+                obs_list.append(obs_arrs[t])
+                act_list.append(act_arrs[t - 1])
+
+            self.assertTrue("context_next_obs" not in for_action)
+            check(for_action["context_obs"], np.stack(obs_list)[None])
+            check(for_action["context_act"], np.stack(act_list[:-1])[None])
```
Comment (gjoliver) [inline rllib/connectors/tests/test_agent.py:382]: I can't seem to figure out why act_list will have 1 fewer element than obs_list?

File: rllib/connectors/tests/test_agent.py [hunk 13]
```diff
@@ -381,44 +330,133 @@ def test_vr_connector_with_multiple_buffers(self):
...
+            action, state, extra_fetch = policy.compute_actions_from_input_dict(
+                eval_batch
+            )
+            next_obs, rewards, dones, info = env.step(action)
+            env_out_dict = {
+                SampleBatch.NEXT_OBS: next_obs,
+                SampleBatch.REWARDS: rewards,
+                SampleBatch.DONES: dones,
+                SampleBatch.INFO: info,
+                SampleBatch.ACTIONS: action,
```
Comment (gjoliver) [inline rllib/connectors/tests/test_agent.py:448]: ah we are providing ACTIONS column here.
so state_buffer connector is not actually doing anything in this setup, since it's not receiving any policy output callbacks.
it's fine. may be something we will come back and improve later.

File: rllib/evaluation/collectors/agent_collector.py [hunk 14]
```diff
@@ -0,0 +1,545 @@
...
+                    sub_list.append(flattened[i])
+
+        # In inference mode, we don't need to keep all of trajectory in memory
+        # we only need to keep the steps required. We can pop from the beginning to
+        # create room for new data.
+        if not self.training:
+            for k in self.buffers:
+                for sub_list in self.buffers[k]:
+                    if sub_list:
+                        sub_list.pop(0)
```
Comment (gjoliver) [inline rllib/evaluation/collectors/agent_collector.py:224]: we can simply pop(0) because we already built all the shift_before elements in self.buffers, right?

File: rllib/evaluation/collectors/agent_collector.py [hunk 15]
```diff
@@ -0,0 +1,545 @@
...
+                data.append(element_at_t[None])
+
+            if data:
+                batch_data[view_col] = self._unflatten_as_buffer_struct(data, data_col)
+
+        batch = self._get_sample_batch(batch_data)
+        return batch
+
+    # TODO: @kouorsh we don't really need view_requirements anymore since it's already
+    # and atribtue of the class
```
Comment (gjoliver) [inline rllib/evaluation/collectors/agent_collector.py:286]: typo: an attribute of the class.

File: rllib/connectors/agent/view_requirement.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/connectors/agent/view_requirement.py:82]: oh actually it's just a random # sampled from a small range today.
there is an open issue thread against me to make this truly unique.
I was proposing to create some scheme based on worker and vector index to make this id unique across a single run.

File: rllib/evaluation/collectors/simple_list_collector.py (hunk 6 above)
Comment (gjoliver) [reply rllib/evaluation/collectors/simple_list_collector.py]: ok cool. thanks.

**Review by gjoliver — APPROVED** [review]
ok, this looks very good.
I have a couple of minor questions left. but let me stamp now.
let's make sure tests pass and merge.
thanks a ton.

File: rllib/connectors/tests/test_agent.py (hunk 11 above)
Comment (kouroshHakha) [reply rllib/connectors/tests/test_agent.py:371]: you need to do consistent zero padding for the learning part. If we don random padding it will mess up with the neural network that needs to learn that zero means invalid timestep.

File: rllib/connectors/tests/test_agent.py (hunk 12 above)
Comment (kouroshHakha) [reply rllib/connectors/tests/test_agent.py:381]: action goes from -c+1:-1 and obs goes from -c+1:0 in view_requirements.

File: rllib/evaluation/collectors/agent_collector.py (hunk 14 above)
Comment (kouroshHakha) [reply rllib/evaluation/collectors/agent_collector.py:227]: no sub_list can still be empty. I had this bug and then I added the check.

File: rllib/evaluation/collectors/agent_collector.py (hunk 15 above)
Comment (kouroshHakha) [reply rllib/evaluation/collectors/agent_collector.py:291]: done.
